### ISSUE TITLE
AttributeError: 'Options' object has no attribute '_fill_fields_cache'

### ISSUE THREAD

**Issue by mr-linch** [deschler/django-modeltranslation#304]
Hello, guys!

Thax for package. 

I install django-modeltranslation, and register some models. When I run `./manage.py makemigrations` I get:

```
Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/home/linch/.pyenv/versions/watch_face_market/lib/python3.4/site-packages/django/core/management/__init__.py", line 338, in execute_from_command_line
    utility.execute()
  File "/home/linch/.pyenv/versions/watch_face_market/lib/python3.4/site-packages/django/core/management/__init__.py", line 312, in execute
    django.setup()
  File "/home/linch/.pyenv/versions/watch_face_market/lib/python3.4/site-packages/django/__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/home/linch/.pyenv/versions/watch_face_market/lib/python3.4/site-packages/django/apps/registry.py", line 115, in populate
    app_config.ready()
  File "/home/linch/.pyenv/versions/watch_face_market/lib/python3.4/site-packages/modeltranslation/apps.py", line 11, in ready
    handle_translation_registrations()
  File "/home/linch/.pyenv/versions/watch_face_market/lib/python3.4/site-packages/modeltranslation/models.py", line 81, in handle_translation_registrations
    autodiscover()
  File "/home/linch/.pyenv/versions/watch_face_market/lib/python3.4/site-packages/modeltranslation/models.py", line 32, in autodiscover
    import_module(module)
  File "/home/linch/.pyenv/versions/3.4.2/lib/python3.4/importlib/__init__.py", line 109, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 2254, in _gcd_import
  File "<frozen importlib._bootstrap>", line 2237, in _find_and_load
  File "<frozen importlib._bootstrap>", line 2226, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1129, in _exec
  File "<frozen importlib._bootstrap>", line 1471, in exec_module
  File "<frozen importlib._bootstrap>", line 321, in _call_with_frames_removed
  File "/home/linch/Projects/watch_face_market/faces/translation.py", line 19, in <module>
    translator.register(Face, FaceTranslationOptions)
  File "/home/linch/.pyenv/versions/watch_face_market/lib/python3.4/site-packages/modeltranslation/translator.py", line 416, in register
    add_translation_fields(model, opts)
  File "/home/linch/.pyenv/versions/watch_face_market/lib/python3.4/site-packages/modeltranslation/translator.py", line 153, in add_translation_fields
    model._meta._fill_fields_cache()
AttributeError: 'Options' object has no attribute '_fill_fields_cache' 
```

Python: 3.4.2
Django: 1.8

**Comment by DanielLoeb** [comment]
The internal django function _fill_fields_cache() was removed in the following commit: https://github.com/django/django/commit/fb48eb05816b1ac87d58696cdfe48be18c901f16

This commit was a Google Summer of Code project, wherein the entire API for the retrieval of fields from Model._meta was rewritten. So I'd expect that, while this speciffic error ticket refers to a single occurence of _fill_fields_cache in translator.py, other parts of django-modeltranslation might also be affected.

I will try to look further into the issue, but IMO it would be best if somebody who (unlike me) is already familiar with the code in question could also have a look at it.

**Comment by DanielLoeb** [comment]
Ok. So here's the current state of affairs: 

I replaced the line 

> model._meta._fill_fields_cache()

in modeltranslation/translator.py with the recommended

> model._meta.get_fields()

which allowed me to run ./migrate.py sync_translation_fields and ./migrate.py makemigrations.

However, as expected, things aren't so easy. When running my project, I now get errors that can be traced back to the line containing

> get_translatable_fields_for_model(f.related_model)

in modeltranslation/manager.py. Here, f.related_model should return a class, but returns a string object containing "generic.AssignedKeyword" instead.

This looks to me as if a migration was borked, but I'm a total database noob, so I might be wrong.

The problem here is that I have no way of knowing if this borkage is actually related to the _fill_fields_cache problem or might be something that is specific to my project.

So there are two questions that I'd like to have answered:

a) Does anybody have a minimal example project for this ticket that produces the _fill_fields_cache error?  This would allow me to check if these new errors are actually caused by my fix attempt or if they are unrelated.

b) Does anybody have a clue why f.related_model would return a string object?

**Comment by DanielLoeb** [comment]
On further inspection, I've found something to indicate I'm stil dealing with the fallout of the _fil_fields_cache issue.

The Docs say that I should run ./manage.py update_translation_fields because I'm dealing with a pre-existing project that was already in use before incorporating django-modeltranslation. 

When I run that command, it successfully updates a bunch of models and then hangs on this:

> Updating data of model '\<class 'mezzanine.pages.models.Link'\>'
> Traceback (most recent call last):
>  File "/var/www/virtualenv-django18/lib/python3.4/site-packages/django/db/models/options.py", line 554, in get_field
>    return self.fields_map[field_name]
>KeyError: 'titles_en'

I've checked what keys are contained in self.fields_map when that happens:

> dict_keys(['children'])

I'll investigate further. Any ideas are welcome.

**Comment by DanielLoeb** [comment]
Now here is something interesting. I've taken a look at the tables that were successfully updated by update_translation_fields.

Turns out, every single one of them has text fields that were replaced by _de and _en versions. The first one which does not have any text fields (in fact only consists of a single integer column) is pages_link, the very table that causes the migration to crash...

**Comment by DanielLoeb** [comment]
So I've got to do some other stuff now, thus a short description of the current state of affairs:

django.db.models.Options._forward_fields_map does not contain the _de and _en modified fields for the Mezzanine model "link", whose database only contains a column for a page_ptr_id, which I guess is a pointer to the database entry of the corresponding "page" (since the Mezzanine class "link" is derived from "page").

I'll continue digging at some later point in time.

**Comment by Kniyl** [comment]
@DanielLoeb What about using
```
model._meta._expire_cache()
model._meta.get_fields()
```
in replacement of ```model._meta._fill_fields_cache()``` to mimic the actual behavior of refilling the cache.

**Comment by Kniyl** [comment]
Spent the last couple of hours testing it with a very simple project. Seems to do the trick. You can get the patch at https://gist.github.com/Kniyl/a515bc600fb4b949da85

**Comment by DanielLoeb** [comment]
@Kniyl

Thank you. It would have taken me a long time of rummaging around to find that out.

Using 

> model._meta._expire_cache()
> model._meta.get_fields()

as a replacement has completely resolved the issues with ./manage.py update_translation_fields.py. 

Running 

> ./manage.py sync_translation_fields
> ./manage.py update_translation_fields
> ./manage.py makemigrations

completes without error and a look into the database shows that all the modified model's tables had their corresponding _de and _en columns added.

So I would be tempted to say this error is fixed and the ticket can be closed.

Unfortunately one issue remains:

>    get_translatable_fields_for_model(f.related_model)

> in modeltranslation/manager.py. Here, f.related_model should return a class, but returns a string object containing "generic.AssignedKeyword" instead.

I am very much aware that this could be independent of the issue of this ticket. Still, I'd like to take a look at the two functions to find out if there is a connection. I will report back here within a day or so with my findings (or lack thereof) so we can hopefully close this ticket.

**Comment by Kniyl** [comment]
Odd…

When I manually check "relation" fields (such as OneToOne or ForeignKey) returned by the ```get_fields()``` call, their ```related_model``` attribute is actually a class.

Are you able to provide the models you are testing with? I'd like to try to reproduce the error.

**Comment by stephenmcd** [comment]
The `generic.AssignedKeyword` is all throughout Mezzanine - it's on the `Page` and `BlogPost` models and probably others.

**Comment by stephenmcd** [comment]
Here's some notes on it: http://mezzanine.jupo.org/docs/utilities.html#keywords

**Comment by Kniyl** [comment]
Damned. I should have thought of using it instead of plain old Django.
But as far as _fill_fields_cache is concerned, the issue seems cornered.

**Comment by DanielLoeb** [comment]
So far I have not found any indication that the current problem is related to the _fill_fields_cache issue.

I could open a new issue for that problem in the morning, if there is no opposition to that.

In any case, here's some possibly relevant information: 

For most fields of Mezzanine models,  f.related_model in get_fields_to_translatable_models(model) is really a class, as it should be. A string object containing "generic.AssignedKeyword" (which is the default string reference) is returned for f.related_model only in cases when f is the KeywordsField "pages.Page.keywords". It might be a coincidence, but so far I've only seen the string object for that particular f.

The general circumstances are a Mezzanine project with regular pages, a Blog and a Cartridge shop. Before the above-mentioned migration was done, there were pages, blogposts and shop sites filled with data. There were regular sites and blog posts both with keywords and without keywords.

So a rather generic case.

**Comment by stephenmcd** [comment]
Thanks for the merge @deschler - with that fixed, we're now just waiting for a modeltranslation release to PyPI before we can release the new version of Mezzanine with modeltranslation support. Sorry for nagging!

**Comment by deschler** [comment]
No problem, wanted to do a bugfix release this week anyway. 0.9.1 is available in pypi now. Thanks to all!

**Comment by stephenmcd** [comment]
Awesome! Thanks Dirk.

**Comment by alfonsoperez** [comment]
Thank you! +1:

**Comment by stratoukos** [comment]
This issue is not fixed for proxy models. 

The reason is that `model._meta._expire_cache()` is only called in `add_translation_fields` which is [never called for proxy models](https://github.com/deschler/django-modeltranslation/blob/3df9d3e5d3f2fc092168789cbcab00c3942ae9fa/modeltranslation/translator.py#L418). Calling `model._meta._expire_cache()` inside `delete_cache_fields` seems to fix it for proxy models as well. 

Calling `model._meta.get_fields()` is not really required, since the cache will be populated whenever someone accesses something that ends up in `get_fields` (like `prepopulated_fields`).

I can open a separate issue (and a PR) if it would help.

Thanks!

PS. Until this is fixed, you can work around it in your project by calling `MyProxyModel._meta._expire_cache()` immediately after registering `MyProxyModel` with modeltranslate.

**Comment by deschler** [comment]
@stratoukos Yes, please create a pull request.
